File: Algorithm_I_study_plan/middle_of_llist.cc

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        if (!head || !head->next) return head;
        ListNode *slow = head;
        ListNode *fast = head;
        while (1) {
            if (fast->next && fast->next->next) {
                slow = slow->next;
                fast = fast->next->next;
            } else if (!fast->next) {
                return slow;
            } else {
                return slow->next;
            }
        }
    }
};

File: Algorithm_I_study_plan/middle_of_llist.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        l = head
        r = head
        
        if head == None or head.next == None:
            return head
        
        while r.next != None and r.next.next != None:
            l = l.next
            r = r.next.next
            
        if r.next == None:
            return l
        
        return l.next
        